### PR TITLE
Fix outstanding JS lint issues

### PR DESCRIPTION
**Pull request by patcon**
Resolves: https://github.com/GCTC-NTGC/gc-digital-talent/issues/2043

Depends on: https://github.com/GCTC-NTGC/gc-digital-talent/pull/1993 (requires merge first)

## To Do

- [x] auto-fix all in frontend/*
- [x] manually fix errors: frontend/common
- [x] manually fix errors: frontend/admin
- [x] manually fix errors: frontend/talentsearch
- [x] enable failures on lint workflow https://github.com/GCTC-NTGC/gc-digital-talent/pull/1993
- [x] manually fix warnings in frontend
- interruption
  - [x] consolidate editorconfig files and resolve misconfiguration
  - [x] migrated eslint from json to yaml for improving readability/comments
  - [x] filed upstream ~~bugfix~~ bug _reproduction_ and added work-around https://github.com/eslint/eslint/issues/15617
  - [x] converted eslint from JSON to YAML (legibility and comments)
  - [x] consolidated all 3 frontend eslint config files into one
  - [x] enable rule to convert all `any` types to `unknown`
  - [x] figure out which commit storybook stopped working. ==> first bad = ef5a3f2504e72647af0526a6af093672e65f7a4a
    - [x] write test script https://gist.github.com/patcon/5cd649196e113ab74cf30255ca8164e7#file-2-bisect-test-sh
    - [x] run git-bisect https://gist.github.com/patcon/5cd649196e113ab74cf30255ca8164e7
  - [x] ignore rule that's leading to issue (any=>unknown)
  - [x] ~~fix issue with new `unknown` types at compile-time~~ `wont-do`
  - [x] added gherkin tests for manual QA 5a2117a5ae9fd532cc7276a5ef34b58b6aae3420
  - [ ] figure out where failure starts happening
- spin-off issues
  - [ ] migrate "any" types to "unknown"s
  - [ ] figure out `import/prefer-default-export` rule
  - [ ] fail linting on all warnings? https://github.com/GCTC-NTGC/gc-digital-talent/pull/2082#discussion_r810774709
  - [x] fix debounce usage with usecallback https://github.com/GCTC-NTGC/gc-digital-talent/pull/2082/files#diff-30ab133bdf859b4d8f08fdd914329e912c392b01e818ff69f3714914258058caR144
      - https://github.com/GCTC-NTGC/gc-digital-talent/pull/2082/files#r827077737
      - https://github.com/GCTC-NTGC/gc-digital-talent/issues/2256

### CODE REVIEW

File: frontend/common/.eslintrc.yml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  '@typescript-eslint/no-shadow': error
+  react/function-component-definition: 'off'
+  # The addition of this rule is a temporary work-around.
+  # TODO: Remove after the relevant fix is merged.
+  # See: https://github.com/eslint/eslint/issues/15617
+  no-restricted-exports:
+    - error
+    - restrictedNamedExports:
+      #- default
+      - then
```
Comment (patcon) [inline frontend/common/.eslintrc.yml:109]: Specifically this was added beyond a straight convert.

**Comment by patcon** [comment]
Sorry @petertgiles, misspoke -- filed bug and "minimal bug reproduction", not bugfix (yet? on my own time perhaps)

**Comment by petertgiles** [comment]
> Sorry @petertgiles, misspoke

No, this is cool.  I was wondering about that linting error assuming I just didn't understand how it was supposed to work.  Cool to see the discussion in the other thread.

File: frontend/.eslintrc.yml [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  '@typescript-eslint/no-empty-function': warn
+  no-use-before-define: 'off'
+  '@typescript-eslint/ban-types': warn
+  '@typescript-eslint/no-use-before-define': warn
+  no-shadow: 'off'
+  '@typescript-eslint/no-shadow': error
+  '@typescript-eslint/no-explicit-any':
+    - warn
+    # See: https://dmitripavlutin.com/typescript-unknown-vs-any/
+    - fixToUnknown: true
```
Comment (patcon) [inline frontend/.eslintrc.yml:106]: Added this to allow `npm run lint:fix` to change all `any` to `unknown`

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -37,11 +37,10 @@ jobs:
...
+        run: |
+          npm run codegen --workspaces --if-present
+          npm run intl-compile --workspaces --if-present
 
       - name: "Run ESlint: all workspaces"
         working-directory: frontend
-        # Temporarily ignore until we fix existing issues.
-        # See: https://github.com/GCTC-NTGC/gc-digital-talent/issues/2043
-        continue-on-error: true
         run: npm run lint --workspaces --if-present
```
Comment (patcon) [inline .github/workflows/lint.yml:46]: If we want `npm run lint` to fail if there are even any warnings, we can use this, which passed the flag through `npm run` down to `eslint`.

```suggestion
        run: npm run lint --workspaces --if-present -- --max-warnings=0
```

File: frontend/admin/resources/js/types/react-table-config.d.ts [hunk 4]
```diff
@@ -121,7 +121,7 @@ declare module "react-table" {
 
   export interface Cell<
     D extends Record<string, unknown> = Record<string, unknown>,
-    V = any,
+    V = unknown,
```
Comment (patcon) [inline frontend/admin/resources/js/types/react-table-config.d.ts:124]: Not sure why this is here, but related to react-table. To ignore

```suggestion
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    V = unknown,
```
```suggestion
    V = unknown,
```

File: frontend/common/src/types/utilityTypes.ts [hunk 5]
```diff
@@ -57,22 +57,3 @@ export type FlattenNested<T> = {
...
-      happiness: string;
-      children: string[];
-    };
-  };
-};
-
-type paths = NestedPaths<test>;
-type years = NestedPropType<test, "parents.marriage.years">;
-type fromArray = FromArray<string>;
-type flattened = FlattenNested<test>;
```
Comment (patcon) [inline frontend/common/src/types/utilityTypes.ts:78]: All unused

File: frontend/talentsearch/resources/js/components/search/SearchForm.tsx [hunk 6]
```diff
@@ -141,14 +141,13 @@ export const SearchForm: React.FunctionComponent<SearchFormProps> = ({
 
   // Whenever form values change (with some debounce allowance), call updateCandidateFilter
   const formValues = watch();
-  const submitDebounced = useCallback(
+  const submitDebounced = useCallback(() => {
```
Comment (patcon) [inline frontend/talentsearch/resources/js/components/search/SearchForm.tsx:144]: This was required by an obscure rule. Apparently useCallback need an inline function:
https://reactjs.org/docs/hooks-reference.html#usecallback

File: frontend/talentsearch/resources/js/tests/skills/SkillResults.test.tsx [hunk 7]
```diff
@@ -17,8 +17,12 @@ function renderSkillResults(title: string) {
       addedSkills={[skills[0], skills[1]]}
       skills={skills}
       title={title}
-      handleAddSkill={async () => {}}
-      handleRemoveSkill={async () => {}}
+      handleAddSkill={async () => {
+        // do nothing.
+      }}
```
Comment (patcon) [inline frontend/talentsearch/resources/js/tests/skills/SkillResults.test.tsx:22]: No empty functions rule

File: frontend/.eslintrc.yml [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  - 'plugin:@typescript-eslint/recommended'
+  - 'plugin:prettier/recommended'
+  - prettier
+globals:
+  Atomics: readonly
+  SharedArrayBuffer: readonly
+parserOptions:
+  ecmaFeatures:
+    jsx: true
+  ecmaVersion: 2020
```
Comment (patcon) [inline frontend/.eslintrc.yml:26]: Standardized all on 2020 (some were 2018)

File: frontend/admin/resources/js/types/react-table-config.d.ts [hunk 9]
```diff
@@ -121,7 +121,8 @@ declare module "react-table" {
 
   export interface Cell<
     D extends Record<string, unknown> = Record<string, unknown>,
-    V = any,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    V = unknown,
```
Comment (patcon) [inline frontend/admin/resources/js/types/react-table-config.d.ts:125]: Not sure how interfaces are working, and whether this is false positive or something real.

**Comment by patcon** [comment]
Ready for review again. When I consolidated the eslint config, a bunch more errors popped up.

I dealt with all the warnings (called attention to the things that might catch your attention).

There's a suggested change above to make it so that all warnings fail the lint test. This might be annoying at first, but strikes me as maybe worthwhile. Might be simpler to correct if someone implements https://github.com/GCTC-NTGC/gc-digital-talent/issues/1869

**Comment by patcon** [comment]
Wrote throwaway git-bisect script, and ran it over lunch... Figured out the first bad commit! https://gist.github.com/patcon/5cd649196e113ab74cf30255ca8164e7

**Comment by patcon** [comment]
Ugh! Fixing this final errors in the morning :/
https://github.com/GCTC-NTGC/gc-digital-talent/runs/5547789383?check_suite_focus=true#step:9:33

File: frontend/talentsearch/resources/js/components/search/SearchForm.tsx (hunk 6 above)
Comment (patcon) [reply frontend/talentsearch/resources/js/components/search/SearchForm.tsx:144]: I'm actually having issues with this one, specifically where it's called below context in L155 and L159

Can anyone advise on how debouncing should work here? It's new to me :)

The examples of can find for addressing the rule use [`debounce-fn`](https://github.com/sindresorhus/debounce-fn) pkg, and we use [`debouce`](https://github.com/component/debounce)

cc @tristan-orourke as author

File: frontend/common/.eslintrc.yml (hunk 1 above)
Comment (patcon) [reply frontend/common/.eslintrc.yml]: was just FYI. no action

File: frontend/.eslintrc.yml (hunk 8 above)
Comment (patcon) [reply frontend/.eslintrc.yml:26]: fyi only. no needed actions.

File: frontend/talentsearch/resources/js/tests/skills/SkillResults.test.tsx (hunk 7 above)
Comment (patcon) [reply frontend/talentsearch/resources/js/tests/skills/SkillResults.test.tsx]: fyi only. no needed actions.

File: frontend/common/src/types/utilityTypes.ts (hunk 5 above)
Comment (patcon) [reply frontend/common/src/types/utilityTypes.ts:78]: fyi only. no needed actions.

File: .github/workflows/lint.yml (hunk 3 above)
Comment (patcon) [reply .github/workflows/lint.yml:46]: Added to todos to spin out another issue. no action here

File: frontend/admin/resources/js/types/react-table-config.d.ts (hunk 9 above)
Comment (patcon) [reply frontend/admin/resources/js/types/react-table-config.d.ts]: fyi only. no needed actions.

**Review by petertgiles — CHANGES_REQUESTED** [review]
Sorry, Pat, I think there are some things that are no bueno.  On the admin site I'm getting the "tab is unresponsive error" and on the talent site my scrolling is stuck to the top of the page.  Are these things happening for you?

**Comment by patcon** [comment]
> Sorry, Pat, I think there are some things that are no bueno. On the admin site I'm getting the "tab is unresponsive" error and on the talent site my scrolling is stuck to the top of the page. Are these things happening for you?

@petertgiles Ah drats. I don't even know all the corners of the app yet, so I admit I didn't do QA testing. Plus I didn't think this would affect anything. Alas...

Might you be able to confirm reproductions in general BDD(ish) format, and I'll do a git-bisect sorta deal to figure out where things went wrong? Great simple docs here: https://cucumber.io/docs/gherkin/reference/

We could use anything we write here to kinda kick-off the version-controlling of some manual QA tests. Plus, if we ever get e2e test framework in place, we can maybe convert these to executable 🎉 

My stubs that I could use your help with to be absolutely clear what you were seeing :)

---

### Admin: Tab issue
```
Scenario: Changing tabs
  Given the following page is loaded: /some/admin/path
  And I try to click the tab "xxxxx"
  Then the tab loads
```

### Talent: scrolling
```
Scenario: Scroll on a page
  Given the following page is loaded: /some/talent/path
  And I try to scroll down with the mouse wheel
  Then the page scrolls down
```

**Comment by petertgiles** [comment]
I'm afraid I don't know any BDD (yet!) but here's my general "Is the app still working?" test when I'm doing a review:

Setup
1. Check out the branch
2. Run the refresh_all script

Talent
1. Navigate to /talent
2. Expect that a non-zero candidate estimate comes up
3. Scroll to the bottom and click _Request Candidates_
4. Fill in junk data to the request form and click _Submit Request_
5. Expect the _Request created successfully!_ toast.

Admin
1. Navigate to /admin
2. If I'm logged in then click _Logout_
3. Click _Login_
4. Follow through the login procedure
5. Click the _Users_ link and expect to see the list of users
6. Click the _Edit_ link beside the first one
7. Edit the first name of the user and click _Submit_
8. Expect to see the edited name showing in the list

It's manual testing :-1: but I feel like I can catch about 90% of errors in about 2 minutes :+1:

**Comment by brindasasi** [comment]
I see the same issue as talent/search stuck in page top. can't scroll down

**Pull request closed without merging**